Refactoring and cleaning up.

qingling128 opened the pull request:
* Remove unnecessary instance variables: `@running_on_managed_vm`,
  `@gae_backend_name` and `@gae_backend_version`.
* Refactor the `configure` function to improve readability by separating
  different sections into smaller functions.
* Separate the logic to set labels for different monitored resources to improve
  readability and maintainability.
* Add comments for confusing logic in the code and explain the story behind them
  to improve readability and also serve as a reminder that this might be
  something we could revisit.
* Combine tests.

igorpeshansky commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -36,8 +36,10 @@ class Map
   # fluentd output plugin for the Stackdriver Logging API
   class GoogleCloudOutput < BufferedOutput
-    # Constants for service names and resource types.
+    # Constants used by the plugin and its tests.
```
s/used/shared/? But I fail to see the reason behind this change...
How about renaming this module to `ServiceConstants` and leaving the comment as-is?

qingling128 replied:
Oops, this change is not part of the refactoring. It only makes sense after adding Metadata Agent support. I just reverted it.

igorpeshansky commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -207,6 +206,24 @@ def initialize
     end
 
+    # Set up regexp patterns to parse tags and logs.
+    def setup_regexp_patterns
```
Why did you add this before `configure`, when all of the other new functions are after? Does this function even need to be public?

qingling128 replied:
Oh, I forgot to move this to where it should belong. Just moved. Also rename it to be `set_regexp_patterns` instead of `setup_regexp_patterns`.

igorpeshansky commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -489,12 +431,14 @@ def extract_entry_labels(group_resource, record)
 
+      # GKE containers.
       if group_resource.type == CONTAINER_CONSTANTS[:resource_type]
         # Move the stdout/stderr annotation from the record into a label
```
Since you're cleaning up, might as well add the missing trailing "." here...

qingling128 replied:
Added.

igorpeshansky commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -225,160 +242,47 @@ def configure(conf)
-      end
+      # Retrieve monitored resource.
+      #
+      # TODO(qingling128) After Metadata Agent support is added, try retrieving
```
Did you mean to include this change?

qingling128 replied:
Yeah, I'm trying to give a hint why the naming is `via_legacy`.

igorpeshansky commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -225,160 +242,47 @@ def configure(conf)
+      #
+      # TODO(qingling128) Revisit this after Metadata Agent supports Cloud
+      # Functions or after we support Cloud Functions in more platforms
+      # (Right now we only supports GKE).
```
Nit: s/supports/support/.

qingling128 replied:
This comment is removed.

igorpeshansky commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -225,160 +242,47 @@ def configure(conf)
+      # we have determined resource type. The purpose is to avoid repeated calls
+      # to metadata server.
+      #
+      # TODO(qingling128) Revisit this after Metadata Agent supports Cloud
```
Did you mean to include this change?

qingling128 replied:
Removed.

igorpeshansky commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -225,160 +242,47 @@ def configure(conf)
-          @project_id.nil?
-      end
+      # Set variables specific to CLoud Functions. This has to be called after
+      # we have determined resource type. The purpose is to avoid repeated calls
```
Nit: the resource type.

qingling128 replied:
Done.

igorpeshansky commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -225,160 +242,47 @@ def configure(conf)
+      # TODO(qingling128) Revisit this after Metadata Agent supports Cloud
+      # Functions or after we support Cloud Functions in more platforms
+      # (Right now we only supports GKE).
+      set_cloudfunctions_variables
```
Can this be folded into another `set_*_variables` function (to be called here)? I'm just wondering why Cloud Functions have to be treated specially.

qingling128 replied:
I thought about it. Yet this is the only variable that has to be set after resource_type is determined. Maybe in the future when other resource types need to set certain variables as well, we can group them. Another option is to fold this into `determine_agent_level_common_labels` in a `Cloud Functions` section. But I leaned towards keeping them separate.

igorpeshansky commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -225,160 +242,47 @@ def configure(conf)
+      #
+      # TODO(qingling128) After Metadata Agent support is added, try retrieving
+      # the monitored resource from Metadata Agent first.
+      @resource = determine_agent_level_monitored_resource_via_legacy
```
Should we combine this with `determine_agent_level_common_labels`? I vaguely recall that you had a function that returned a pair -- what happened to that?

qingling128 replied:
I put these in two separate files because `determine_agent_level_monitored_resource_via_legacy` should be equivalent to `determine_agent_level_monitored_resource_via_metadata_agent` which is to be added soon (in the other PR).

The functions that return a pair of `resource_labels` and `common_labels` are at the `group` and `entry` level.

Note that at the `agent` level we don't need to determine the monitored resource type before calling Metadata Agent. But at `group` and `entry` levels we need to determine monitored resource type locally based on tags first, then call Metadata Agent if necessary (only when we detect special resource type).

igorpeshansky commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -512,17 +456,20 @@ def extract_entry_labels(group_resource, record)
       common_labels.merge!(fields_to_labels(record, @label_map))
 
+      # Cloud Functions.
+      # TODO TODOTODO
```
???

qingling128 replied:
Oh, this is a reminder for myself to double check the influence and move this part to be right after the Cloud Functions tag parsing above. Completed the TODO.

igorpeshansky commented:
Just to clarify, I'm only asking for separate commits in this PR, not separate PRs.

qingling128 commented:
I've split the PR into different commits. I did not do a force push because I'm not sure how GitHub will react to that. So right now an ugly revert commit is present. I will make sure the commit message is clean when I merge. When merging, do we want to just squash everything into one single commit or keep different sections in different commits?

qingling128 commented:
PTAL.

qingling128 commented:
Friendly ping.

qingling128 commented on test/plugin/base_test.rb:
```diff
@@ -701,80 +701,33 @@ def test_multiple_managed_vm_logs
-  end
-
-  def test_multiple_container_logs_metadata_from_tag
+  def test_container_log_metadata
```
This has been split into two tests as suggested in the other comment.

qingling128 commented on test/plugin/base_test.rb:
```diff
@@ -701,80 +701,33 @@ def test_multiple_managed_vm_logs
-        assert_equal CONTAINER_NANOS, entry['timestamp']['nanos'], entry
-        assert_equal CONTAINER_SEVERITY, entry['severity'], entry
+    {
+      # Metadata from metadata.
```
Just clarified this comment. The metadata is in fact from the record:
```
# Test container logs when metadata is extracted from the 'kubernetes' field
# in the log record.
```

qingling128 commented on test/plugin/base_test.rb:
```diff
@@ -701,80 +701,33 @@ def test_multiple_managed_vm_logs
+        CONTAINER_FROM_METADATA_PARAMS,
+      # Metadata from tag.
+      method(:container_log_entry) => CONTAINER_FROM_TAG_PARAMS
+    }.each do |log_entry_method, expected_params|
```
Good point. Just refactored this into two methods with shared utility function.

qingling128 commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -225,160 +222,43 @@ def configure(conf)
-      # TODO: Send instance tags as labels as well?
-      @common_labels = {}
-      @common_labels.merge!(@labels) if @labels
+      # Set regex patterns that is used to parse tags and logs.
```
Changed this one and all other occurrences.

qingling128 replied:
Changed.

qingling128 replied:
Right. Removed.

qingling128 commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -225,160 +222,43 @@ def configure(conf)
-      end
+      # EC2 Metadata server returns everything in one call. Store it to avoid
+      # making multiple calls.
+      @ec2_metadata = fetch_ec2_metadata if @platform == Platform::EC2
```
Good idea. Changed. BTW, I combined `ec2_metadata` and `fetch_ec2_metadata`.

qingling128 commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -225,160 +222,43 @@ def configure(conf)
+      # no longer need to require all these metadata in logging agent. But for
+      # now, they are still required.
+      #
+      # TODO(qingling128) After Metadata Agent support is added, try extracting
```
Added.

qingling128 commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -225,160 +222,43 @@ def configure(conf)
+      #
+      # TODO(qingling128) After Metadata Agent support is added, try extracting
+      # these info from responses from Metadata Agent first.
+      set_required_metadata_variables
```
Yes. The current workflow with metadata agent support added is:

1. Get monitored resource from metadata agent.
2. Set these required variables: `project_id`, `vm_id`, `vm_name`, `zone`, etc. In this step, we will make use of the monitored resource retrieved from metadata agent if available. Only when a variable is not available from the metadata agent response will we try to retrieve it locally.
3. Execute the legacy monitored resource construction code which will make use of the variables that have been set in step 2.

qingling128 commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -764,6 +520,24 @@ def parse_json_or_nil(input)
       nil
     end
 
+    # Set regexp patterns to parse tags and logs.
```
Moved.

qingling128 commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -817,6 +591,409 @@ def fetch_ec2_metadata
+
+    def set_project_id
+      # 1. Check if the value is explicitly in the config thus already set.
+      return unless @project_id.nil?
```
Changed.

qingling128 commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -817,6 +591,409 @@ def fetch_ec2_metadata
+      # 3. Try to obtain it from the credentials.
+      @project_id ||= CredentialsInfo.project_id
+    rescue StandardError => e
+      @log.debug 'Failed to obtain project id: ', error: e
```
Changed this and all other occurrences.

qingling128 commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -817,6 +591,409 @@ def fetch_ec2_metadata
+
+    def set_vm_id
+      # 1. Check if the value is explicitly in the config thus already set.
+      return unless @vm_id.nil?
```
Changed. Also removed the part to extract the value from `@resource`, which is only meaningful after Metadata agent support is added. I left it here by mistake.

qingling128 commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -817,6 +591,409 @@ def fetch_ec2_metadata
+      unless @resource.nil?
+        @zone ||= @resource.labels['location'] if
+          @resource.labels.key?('location')
+        @zone ||= @resource.labels['zone'] if
```
Changed. In fact I removed the entire section to extract the value from `@resource`. This is only meaningful after Metadata agent support is added. I left it here by mistake.

qingling128 commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -817,6 +591,409 @@ def fetch_ec2_metadata
+          @resource.labels.key?('location')
+        @zone ||= @resource.labels['zone'] if
+          @platform == Platform::GCE && @resource.labels.key?('zone')
+        @zone ||= @resource.labels['region'] if
```
Changed. In fact I removed the entire section to extract the value from `@resource`. This is only meaningful after Metadata agent support is added. I left it here by mistake.

qingling128 commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -858,6 +829,189 @@ def set_location
+
+        # Only when @running_cloudfunctions is true will we try to match log
+        # tags against Cloud Functions tag regex when processing log entries.
+        @running_cloudfunctions = true
```
Removed `set_cloudfunctions_variables` and inlined the code.

qingling128 commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -858,6 +829,189 @@ def set_location
+        # tags against Cloud Functions tag regex when processing log entries.
+        @running_cloudfunctions = true
+        # Fetch this info and store it to avoid recurring metadata server calls.
+        @gcf_region = fetch_gce_metadata('instance/attributes/gcf_region')
```
? `def fetch_gce_metadata(metadata_path)`? This method has many occurrences. `fetch_ec2_metadata` has only one occurrence, and I just inlined that.

qingling128 commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -817,6 +591,409 @@ def fetch_ec2_metadata
+      labels
+    end
+
+    # Determin the group level monitored resource and common labels shared by a
```
Oops. Changed.

bmoyles0117 commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -225,160 +222,50 @@ def configure(conf)
-        common_labels["#{COMPUTE_CONSTANTS[:service]}/resource_name"] = @vm_name
+      # We only support Cloud Functions logs for GKE right now.
+      if @resource.type == CONTAINER_CONSTANTS[:resource_type] &&
+         fetch_gce_metadata('instance/attributes/').split.include?('gcf_region')
```
Are the values of fetch_gce_metadata memoized? If not, can we memoize them or store a local value and reuse the value in the if statement to avoid fetching data twice?

qingling128 replied:
Unlike AWS, the GCE metadata server does not return everything in one call. Instead we retrieve different pieces of metadata via different urls. And these metadata are only retrieved once at agent start time. So we do not memorize them.

bmoyles0117 commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -806,15 +569,403 @@ def fetch_gce_metadata(metadata_path)
+        labels["#{COMPUTE_CONSTANTS[:service]}/resource_name"] = @vm_name
+
+      # GKE container.
+      when CONTAINER_CONSTANTS[:resource_type]
```
Can GKE and GCE be blended into a single conditional like Dataflow and Dataproc are?

qingling128 replied:
Sure thing. Just changed this.

bmoyles0117 commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -806,15 +569,403 @@ def fetch_gce_metadata(metadata_path)
+            matched_regex_group['container_name']
+          group_resource_labels['namespace_id'] =
+            matched_regex_group['namespace_name']
+          group_resource_labels['pod_id'] =
```
Should we be consistent with naming across our regex fields and labels?

qingling128 replied:
`namespace_name` and `pod_name` are implicitly mapped to `namespace_id` and `pod_id` per GKE's request. The `namespace_id` and `pod_id` are poorly named. They are in fact names. Yet it's not possible to change it without affecting users at this moment. See details in b/38411250.

qingling128 commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -806,15 +569,403 @@ def fetch_gce_metadata(metadata_path)
+
+    # Set regexp patterns to parse tags and logs.
+    def set_regexp_patterns
+      @compiled_kubernetes_tag_regexp = nil
```
I guess we were trying to differentiate it from the config name `kubernetes_tag_regexp` by prefixing it with `compiled`. At this point, we can not change the config name because that will affect users. `compiled` is not ideal yet kind of makes sense. Any other suggestion?

bmoyles0117 replied:
Are both`@compiled_kubernetes_tag_regexp` necessary here? Will it default to nil anyways if kubernetes_tag_regexp is not set?

qingling128 replied:
Removed the first one.

qingling128 commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -806,15 +569,403 @@ def fetch_gce_metadata(metadata_path)
+      @project_id ||= fetch_gce_metadata('project/project-id') if
+        @platform == Platform::GCE
+      @project_id ||= CredentialsInfo.project_id
+    rescue StandardError => e
```
Yup, that can totally happen. The detailed error and its associated message will be included in the log line though (we pass the error to the logger as well). The only thing we do when there an error occurs is to log it. So I think this should be enough. WDYT?

bmoyles0117 replied:
I'm fine with this as long as the user can distinguish the actual here. SGTM.

qingling128 commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -806,15 +569,403 @@ def fetch_gce_metadata(metadata_path)
+    def set_vm_id
+      @vm_id ||= fetch_gce_metadata('instance/id') if @platform == Platform::GCE
+      @vm_id ||= ec2_metadata['instanceId'] if @platform == Platform::EC2
+    rescue StandardError => e
```
See above.

qingling128 commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -806,15 +569,403 @@ def fetch_gce_metadata(metadata_path)
+        @platform == Platform::GCE
+      @zone ||= 'aws:' + ec2_metadata['availabilityZone'] if
+        @platform == Platform::EC2 && ec2_metadata.key?('availabilityZone')
+    rescue StandardError => e
```
See above.

qingling128 commented:
PTAL

qingling128 commented:
Friendly ping.

bmoyles0117 commented:
LGTM

qingling128 commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -306,11 +306,6 @@ def configure(conf)
 
-      # For each resource type, there is a list of labels that we want to report
-      # as monitored resource instead of metadata labels. Move them if present.
-      @resource.labels.merge!(
```
We are doing the same thing at `agent` level and at `entry` level. The extraction at `agent` level should be redundant because `@label_map` is processed at `entry` level, and we are required to extract these labels after that step anyway. Simply extracting these labels at `entry` level should be sufficient.

qingling128 commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -1357,14 +1367,16 @@ def sanitize_tag(tag)
     # For every original_label => new_label pair in the label_map, delete the
-    # original_label from the record if it exists, and extract the value to form
-    # a map with the new_label as the key.
-    def fields_to_labels(record, label_map)
```
Renamed this function to make it more generic. Also make it obvious what is actually being done.

qingling128 commented:
PTAL.

igorpeshansky commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -873,15 +631,434 @@ def fetch_gce_metadata(metadata_path)
+    def determine_agent_level_common_labels
+      labels = {}
+      # User can specify labels via config. We want to capture those as well.
+      # TODO: Send instance tags as labels as well?
```
Now that we've refactored this code, I don't think this TODO makes sense anymore -- we know that the instance labels will be sent via metadata.

qingling128 replied:
Removed.

igorpeshansky commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -873,15 +631,434 @@ def fetch_gce_metadata(metadata_path)
+      labels.merge!(@labels) if @labels
+
+      case @resource.type
+
```
Same here: this blank line doesn't do anything for readability.

qingling128 replied:
Removed.

igorpeshansky commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -873,15 +631,434 @@ def fetch_gce_metadata(metadata_path)
+
+      # GAE app.
+      when APPENGINE_CONSTANTS[:resource_type]
+        labels["#{COMPUTE_CONSTANTS[:service]}/resource_id"] = @vm_id
```
Let's do `labels.merge!({<hash_literal>})` instead, here and below.

qingling128 replied:
Changed.

igorpeshansky commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -873,15 +631,434 @@ def fetch_gce_metadata(metadata_path)
+        labels["#{EC2_CONSTANTS[:service]}/resource_name"] = @vm_name
+
+      # Cloud ML.
+      when ML_CONSTANTS[:resource_type]
```
This case seems identical to the one for Cloud Dataflow and Cloud Dataproc. Combine?

qingling128 replied:
Combined.

igorpeshansky commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -873,15 +631,434 @@ def fetch_gce_metadata(metadata_path)
+      case @resource.type
+
+      # GAE app.
+      when APPENGINE_CONSTANTS[:resource_type]
```
This case seems identical to the one for Cloud Dataflow and Cloud Dataproc. Combine?

qingling128 replied:
Combined.

igorpeshansky commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -873,15 +631,434 @@ def fetch_gce_metadata(metadata_path)
+        determine_group_level_monitored_resource_type(tag)
+
+      # Determine group level monitored resource labels and common labels.
+      group_resource_labels, group_common_labels = determine_group_level_labels(
```
Let's make this formatting the same as in lines 892-893 above.

qingling128 replied:
Changed.

igorpeshansky commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -873,15 +631,434 @@ def fetch_gce_metadata(metadata_path)
+    # Determine group level monitored resource type shared by a collection of
+    # entries.
+    # Returns the resource type and tag regexp matched groups. The matched
+    # groups only apply to some resource types. Return nil if not applicable or
```
Consistency: you first say `Returns`, and then `Return` in a later sentence.

qingling128 replied:
Changed.

igorpeshansky commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -873,15 +631,434 @@ def fetch_gce_metadata(metadata_path)
+    # if there is no match.
+    def determine_group_level_monitored_resource_type(tag)
+      # Match tag against Cloud Functions format.
+      if @running_cloudfunctions
```
It seems like we could set a single `@tag_regexp` value at the beginning to either `@cloudfunctions_tag_regexp` if `@running_cloudfunctions`, or `@compiled_kubernetes_tag_regexp` if the resource is a container resource, and then this method would become simply:
```ruby
  def determine_group_level_monitored_resource_type(tag)
    matched_regex_group = @tag_regexp ? @tag_regexp.match(tag) : nil
    resource_type = @running_cloudfunctions ? CLOUDFUNCTIONS_CONSTANTS[:resource_type]
                                            : @resource.type
    return [resource_type, matched_regex_group]
  end
```
And with the method becoming this simple, you can even inline it at the call site.

qingling128 replied:
In fact, when `@running_cloudfunctions` is true, we want to match tags against both regexps. So I changed it to be a list instead of a map. Then I realized it's more readable without inlining the function since we need to return early when a regexp is matched.

I also got rid of the `@running_cloudfunctions` variable`.

Initial set up:
```
      # Set regexp that we should match tags against later on. Using a list
      # instead of a map to ensure order. For example, tags will be matched
      # against Cloud Functions first, then GKE.
      @tag_regexp_list = []
      if @resource.type == CONTAINER_CONSTANTS[:resource_type]
        # We only support Cloud Functions logs for GKE right now.
        if fetch_gce_metadata('instance/attributes/'
                             ).split.include?('gcf_region')
          # Fetch this info and store it to avoid recurring
          # metadata server calls.
          @gcf_region = fetch_gce_metadata('instance/attributes/gcf_region')
          @tag_regexp_list << {
            tag_regexp: @cloudfunctions_tag_regexp,
            derived_type: CLOUDFUNCTIONS_CONSTANTS[:resource_type]
          }
        end
        @tag_regexp_list << {
          tag_regexp: @compiled_kubernetes_tag_regexp,
          derived_type: CONTAINER_CONSTANTS[:resource_type]
        }
      end
```

Tag parsing:
```
    def determine_group_level_monitored_resource_type(tag)
      @tag_regexp_list.each do |rule_pair|
        matched_regex_group = rule_pair[:tag_regexp].match(tag)
        return [rule_pair[:derived_type], matched_regex_group] if
          matched_regex_group
      end
      [@resource.type, nil]
    end
```

igorpeshansky commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -873,15 +631,434 @@ def fetch_gce_metadata(metadata_path)
+    # Determine group level monitored resource labels and common labels. These
+    # labels will be shared by a collection of entries.
+    def determine_group_level_labels(group_resource_type,
+                                     matched_regex_group)
```
Definitely fits on the previous line...

qingling128 replied:
Changed.

igorpeshansky commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -873,15 +631,434 @@ def fetch_gce_metadata(metadata_path)
+            matched_regex_group['namespace_name']
+          group_resource_labels['pod_id'] =
+            matched_regex_group['pod_name']
+          %w(namespace_name pod_name).each do |field|
```
This list could also go into `CONTAINER_CONSTANTS`.

qingling128 replied:
Moved this list to `CONTAINER_CONSTANCE[:extra_common_labels]`.

igorpeshansky commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -873,15 +631,434 @@ def fetch_gce_metadata(metadata_path)
+
+      # Cloud Functions.
+      when CLOUDFUNCTIONS_CONSTANTS[:resource_type]
+        group_resource_labels['region'] = @gcf_region
```
Might be more readable as `<labels_var>.merge!({<hash_literal>})`.

qingling128 replied:
Changed.

igorpeshansky commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -873,15 +631,434 @@ def fetch_gce_metadata(metadata_path)
+          decode_cloudfunctions_function_name(
+            matched_regex_group['encoded_function_name'])
+        instance_id = group_resource_labels.delete('instance_id')
+        group_common_labels["#{CONTAINER_CONSTANTS[:service]}/cluster_name"] =
```
This seems reminiscent of `delete_and_extract_labels`...

qingling128 replied:
Seems we can't simply apply that function because we are setting two labels to the same thing `instance_id`.

```
        instance_id = group_resource_labels.delete('instance_id')
        group_common_labels.merge(
          "#{CONTAINER_CONSTANTS[:service]}/instance_id" => instance_id,
          "#{COMPUTE_CONSTANTS[:service]}/resource_id" => instance_id,
...
        )
```

igorpeshansky commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -873,15 +631,434 @@ def fetch_gce_metadata(metadata_path)
+      # GKE container.
+      when CONTAINER_CONSTANTS[:resource_type]
+        if matched_regex_group
+          group_resource_labels['container_name'] =
```
If you don't care about destroying `matched_regex_group`, this also seems reminiscent of `delete_and_extract_labels`...

qingling128 replied:
Changed to use `delete_and_extract_labels`.

igorpeshansky commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -873,15 +631,434 @@ def fetch_gce_metadata(metadata_path)
+        if record.key?('kubernetes')
+          resource_labels.merge!(
+            delete_and_extract_labels(
+              record['kubernetes'], %w(namespace_id pod_id container_name)
```
This list can go into `CONTAINER_CONSTANTS` too (under a different key from the next list :-) ).

qingling128 replied:
Moved this list to `CONTAINER_CONSTANCE[:extra_resource_labels]`.

igorpeshansky commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -873,15 +631,434 @@ def fetch_gce_metadata(metadata_path)
+                .map { |l| [l, "#{CONTAINER_CONSTANTS[:service]}/#{l}"] }.to_h))
+          # Prepend label/ to all user-defined labels' keys.
+          if record['kubernetes'].key?('labels')
+            record['kubernetes']['labels'].each do |key, value|
```
Seems reminiscent of `delete_and_extract_labels` as well.

qingling128 replied:
Changed.

qingling128 commented on test/plugin/base_test.rb:
```diff
@@ -1329,6 +1268,25 @@ def verify_json_log_entries(n, params, payload_type = 'textPayload')
     assert i == n, "Number of entries #{i} does not match expected number #{n}"
   end
 
+  def verify_container_logs(log_entry_method, expected_params)
```
Renamed.

qingling128 commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -873,15 +631,434 @@ def fetch_gce_metadata(metadata_path)
+      set_location
+
+      # All metadata parameters must now be set.
+      return if @project_id && @zone && @vm_id
```
Changed.

qingling128 commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -873,15 +631,434 @@ def fetch_gce_metadata(metadata_path)
+    # Determine agent level monitored resource type.
+    def determine_agent_level_monitored_resource_type
+      # EC2 instance.
+      return EC2_CONSTANTS[:resource_type] if
```
Yup, `case` is more readable. Changed.

qingling128 commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -873,15 +631,434 @@ def fetch_gce_metadata(metadata_path)
-        contents = f.read
-        return JSON.parse(contents)
+
+      # Unknown platform will be defaulted to GCE instance..
```
Removed.

qingling128 commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -873,15 +631,434 @@ def fetch_gce_metadata(metadata_path)
+      # Resource types determined by @subservice_name config.
+      # Cloud Dataflow.
+      return DATAFLOW_CONSTANTS[:resource_type] if
+        @subservice_name == DATAFLOW_CONSTANTS[:service]
```
Changed to use a `SUB_SERVICE_MAP`.

qingling128 commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -873,15 +631,434 @@ def fetch_gce_metadata(metadata_path)
+      labels = {}
+
+      case type
+
```
Removed.

qingling128 commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -873,15 +631,434 @@ def fetch_gce_metadata(metadata_path)
+      case type
+
+      # GAE app.
+      when APPENGINE_CONSTANTS[:resource_type]
```
Changed.

qingling128 commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -873,15 +631,434 @@ def fetch_gce_metadata(metadata_path)
+          labels['version_id'] = fetch_gce_metadata(
+            'instance/attributes/gae_backend_version')
+        rescue StandardError => e
+          @log.error 'Failed to set monitored resource labels for GAE: ',
```
On a second thought, we can just log with the resource type name. So we can achieve having one `begin` and `rescue` without introducing an extra map. PTAL.

qingling128 commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -873,15 +631,434 @@ def fetch_gce_metadata(metadata_path)
+        labels["#{COMPUTE_CONSTANTS[:service]}/resource_name"] = @vm_name
+        labels["#{COMPUTE_CONSTANTS[:service]}/zone"] = @zone
+
+      # GCE and GKE container.
```
Changed.

qingling128 commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -873,15 +631,434 @@ def fetch_gce_metadata(metadata_path)
+      if group_resource.type == DATAFLOW_CONSTANTS[:resource_type]
+        resource_labels.merge!(
+          delete_and_extract_labels(
+            common_labels, %w(region job_name job_id step_id)
```
Moved this list to `DATAFLOW_CONSTANTS[:extra_resource_labels]`.

qingling128 commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -1178,43 +1348,35 @@ def decode_cloudfunctions_function_name(function_name)
-      return {} if label_map.nil? || !label_map.is_a?(Hash)
+    # original_label from the original_resource map if it exists, and extract
+    # the value to form a map with the new_label as the key.
+    def delete_and_extract_labels(original_resource, label_map)
```
Renamed `original_resource` to `hash`.

igorpeshansky commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -873,15 +641,409 @@ def fetch_gce_metadata(metadata_path)
+      when CONTAINER_CONSTANTS[:resource_type]
+        unless matched_regex_group.nil?
+          matched_regex_group_hash = Hash[
+            matched_regex_group.names.zip(matched_regex_group.captures)]
```
Might be worth commenting that we only expect one occurrence of each key in the match group (see [this](https://stackoverflow.com/questions/11688726/convert-named-matches-in-matchdata-to-hash#answer-33837669) for an example of when it would matter).

qingling128 replied:
Interesting. Just added a comment `We only expect one occurrence of each key in the match group.`

igorpeshansky commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -873,15 +641,409 @@ def fetch_gce_metadata(metadata_path)
+
+      # GKE container.
+      when CONTAINER_CONSTANTS[:resource_type]
+        unless matched_regex_group.nil?
```
This could just be `if matched_regex_group`, right?

qingling128 replied:
Oops. Right. Changed.

igorpeshansky commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -873,15 +641,409 @@ def fetch_gce_metadata(metadata_path)
+          group_common_labels.merge!(
+            delete_and_extract_labels(
+              matched_regex_group_hash,
+              CONTAINER_CONSTANTS[:extra_common_labels]
```
[Optional] I find `Hash[list.map { ... }]` easier to read than `list.map { ... }.to_h`, because the hash construction is upfront. Also in other places below.

qingling128 replied:
I was also debating among using `Hash[]`, `to_hash` and `to_h`. After some googling, it appeared to me that `to_h` is more popular. And I did see that more often in Ruby doc examples.

Some discussion around it:
https://stackoverflow.com/questions/28302482/hash-vs-to-h-are-they-the-same-and-if-so-which-is-better
http://zverok.github.io/blog/2016-01-18-implicit-vs-expicit.html

WDYT?

qingling128 replied:
Cool, I'll leave it in the common format that Ruby developers use (`to_h`) then.

igorpeshansky commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -1178,43 +1333,39 @@ def decode_cloudfunctions_function_name(function_name)
-      return {} if label_map.nil? || !label_map.is_a?(Hash)
+    # original_label from the hash map if it exists, and extract the value to
+    # form a map with the new_label as the key.
+    def delete_and_extract_labels(hash, label_map, should_delete = true)
```
Why do you need the `should_delete` flag? In all cases where you set it to `false`, we don't actually care if we destroy the hash.

qingling128 replied:
Ah, I added this because in order to apply `delete_and_extract_labels` to `matched_regex_group`. Right now we are setting matched_regex_group['pod_name'] to both resource label `pod_id` and common label `container.googleapis.com/pod_name`. So we can not destroy the hash on the first run.

igorpeshansky replied:
This seems like an unnecessary complication. Instead, you can either construct the hash before every call to `delete_and_extract_labels`, or just pass in the `.dup` of the hash.

qingling128 replied:
Removed the `should_delete` variable and use `.dup` instead.

qingling128 commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -70,8 +73,10 @@ module Constants
+        resource_type: 'ml_job',
+        extra_common_labels: %w(job_id task_name)
       }
+      SUB_SERVICE_LIST = [DATAFLOW_CONSTANTS, ML_CONSTANTS]
```
Changed.

qingling128 replied:
Good idea. Changed.

qingling128 commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -267,162 +269,52 @@ def configure(conf)
+      # Set regexp that we should match tags against later on. Using a list
+      # instead of a map to ensure order. For example, tags will be matched
+      # against Cloud Functions first, then GKE.
+      @tag_regexp_list = []
```
The original [commit](https://github.com/GoogleCloudPlatform/fluent-plugin-google-cloud/commit/44dccaf724841f90116eb28a1bebf0877a1ed5d7) was from 2 years ago.

```
+        if @running_cloudfunctions
 +          # If the current group of entries is coming from a Cloud Functions
 +          # function, the function name can be extracted from the tag.
 +          match_data = @cloudfunctions_tag_regexp.match(tag)
 +          if match_data
 +            # Service name is set to Cloud Functions only for logs actually
 +            # coming from a function.
 +            @service_name = CLOUDFUNCTIONS_SERVICE
 +            labels = write_log_entries_request['commonLabels']
 +            labels["#{CLOUDFUNCTIONS_SERVICE}/function_name"] =
 +              match_data['function_name']
 +          else
 +            # Other logs are considered as coming from the Container Engine
 +            # service.
 +            @service_name = CONTAINER_SERVICE
 +          end
 +        end
```

`Cloud Functions` is only supported on `GKE`. So we default back to the `GKE` resource type when the tag does not match with `Cloud Function` format. Then we match the tag against `GKE` regex.

I think this makes sense because `@running_cloudfunctions` is true as long as `gcf_region` is available from metadata server. Yet we still want other GKE system logs to be tagged properly.

Also these two tests (`test_multiple_cloudfunctions_logs_tag_not_matched` and `test_multiple_cloudfunctions_logs_tag_not_matched`) were added in this second [commit](https://github.com/GoogleCloudPlatform/fluent-plugin-google-cloud/commit/cb0c64257586a44d50cafc3c1e9ca969c9f1988c) to explicitly ensure the behavior.

WDYT?

qingling128 replied:
No worries.

qingling128 commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -267,162 +269,52 @@ def configure(conf)
+      @tag_regexp_list = []
+      if @resource.type == CONTAINER_CONSTANTS[:resource_type]
+        # We only support Cloud Functions logs for GKE right now.
+        if fetch_gce_metadata('instance/attributes/'
```
I changed `tag_regexp_list` to a list of pairs. Also changed the order of `resource type` and `tag regex`.

As for the error handling change, I found it a bit more confusing after the change. Because catching an error here seems to imply `instance/attributes/gcf_region` is supposed to be present for the most common cases. But in fact it's the opposite. Also, it will make it hard to tell the difference between the `instance/attributes/gcf_region` not being present or the metadata server going down. So I left this part the way it was. Does that make sense?

qingling128 replied:
Cool.

qingling128 commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -267,162 +269,52 @@ def configure(conf)
+          # metadata server calls.
+          @gcf_region = fetch_gce_metadata('instance/attributes/gcf_region')
+          @tag_regexp_list << {
+            tag_regexp: @cloudfunctions_tag_regexp,
```
I think `cloudfunctions_tag_regexp` is already a regex, not a string. We have to do

```
@compiled_kubernetes_tag_regexp = Regexp.new(@kubernetes_tag_regexp)
```

because `@kubernetes_tag_regexp` is a string configurable by users.

qingling128 replied:
Just renamed the variables.

qingling128 commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -873,15 +641,409 @@ def fetch_gce_metadata(metadata_path)
+        end
+
+        # Resource types determined by @detect_subservice config.
+        if @detect_subservice
```
I renamed `SUBSERVICE_MAP` to `SUBSERVICE_NAME_MAP`. And added a `SUBSERVICE_ATTRIBUTE_MAP` to add these items (because they are kind of different). The refactored code is as below.

```
      # The map between a resource type and expected subservice attributes.
      SUBSERVICE_ATTRIBUTE_MAP = {
        # GAE.
        APPENGINE_CONSTANTS[:resource_type] =>
          %w(gae_backend_name gae_backend_version),
        # GKE.
        CONTAINER_CONSTANTS[:resource_type] => %w(kube-env),
        # Cloud Dataproc.
        DATAPROC_CONSTANTS[:resource_type] =>
          %w(dataproc-cluster-uuid dataproc-cluster-name)
      }
```
```
        if @detect_subservice
          begin
            attributes = fetch_gce_metadata('instance/attributes/').split
            SUBSERVICE_ATTRIBUTE_MAP.map do |resource_type, expected_attributes|
              return resource_type if
                expected_attributes.all? { |attr| attributes.include?(attr) }
            end
          rescue StandardError => e
            @log.error 'Failed to detect subservice: ', error: e
          end
        end
```

PTAL.

qingling128 replied:
I renamed the maps to avoid confusion.

Also changed
`return resource_type if expected_attributes.all? { |attr| attributes.include?(attr) }`
to 
`return resource_type if attributes.superset?(expected)`.

I don't get why we should put `service -> attributes` in `SUBSERVICE_METADATA_ATTRIBUTES`, then do another map lookup to get `service -> resource_type` though. Can't we just put `resource_type -> attributes` in `SUBSERVICE_METADATA_ATTRIBUTES`? Ultimately we want to get the resource types.

rbuskens replied:
I agree. service -> resource_type and resource_type -> attributes makes more sense to me.

Naive question: where do you actually use SUBSERVICE_MAP?

qingling128 replied:
That's not naive at all. Turned out I did not rename the variable at the other side. I just changed it. I also created a ticket to add unit tests for logic around `@subservice_name` so this can be captured in the future.

rbuskens replied:
LGTM.

qingling128 commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -873,15 +641,409 @@ def fetch_gce_metadata(metadata_path)
+      case type
+      # GAE app.
+      when APPENGINE_CONSTANTS[:resource_type]
+        labels = {
```
Changed.

igorpeshansky commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -873,15 +656,397 @@ def fetch_gce_metadata(metadata_path)
+              matched_regex_group_hash, {
+                # The kubernetes_tag_regexp is poorly named. 'namespace_name' is
+                # in fact 'namespace_id'. 'pod_name' is in fact 'pod_id'.
+                'container_name' => 'container_name',
```
The keys in this hash are `CONTAINER_CONSTANTS[:extra_resource_labels]`. Let's add a TODO to figure out how to reuse them.

qingling128 replied:
Added.

qingling128 commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -873,15 +656,397 @@ def fetch_gce_metadata(metadata_path)
+      # GAE app.
+      when APPENGINE_CONSTANTS[:resource_type]
+        return {
+          'module_id' => fetch_gce_metadata(
```
Changed.

qingling128 commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -873,15 +656,397 @@ def fetch_gce_metadata(metadata_path)
+        return labels
+      end
+
+      return {}
```
Seems it's fine with this format. Yet I just changed to `{}` anyway.

qingling128 commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -873,15 +656,397 @@ def fetch_gce_metadata(metadata_path)
+        if @detect_subservice
+          begin
+            attributes = fetch_gce_metadata('instance/attributes/').split
+            SUBSERVICE_ATTRIBUTE_MAP.map do |resource_type, expected_attributes|
```
Ah, right. `map` is an overkill here. Changed to `each`.

qingling128 commented:
I've addressed all the feedback. Yet seems some of my comments did not show up in the main comment history (probably because I rebased the branch and Github got confused). But if you click on the actual commit, that view should show all the commits. PTAL.

bmoyles0117 commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -70,7 +73,28 @@ module Constants
+      SUBSERVICE_MAP.default = COMPUTE_CONSTANTS[:resource_type]
+
+      # The map between a resource type and expected subservice attributes.
+      SUBSERVICE_METADATA_ATTRIBUTES = {
```
Could/should we define metadata attributes as a property of the constants and define this map the same way we define the SUBSERVICE_MAP?

qingling128 replied:
Changed.

bmoyles0117 commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -873,15 +659,399 @@ def fetch_gce_metadata(metadata_path)
+      # GKE container.
+      when CONTAINER_CONSTANTS[:resource_type]
+        raw_kube_env = fetch_gce_metadata('instance/attributes/kube-env')
+        kube_env = YAML.load(raw_kube_env)
```
Could kube_env end up being defined as nil in an error situation? If so, I think `cluster_name_from_kube_env` might blow up attempting to access the key.

qingling128 replied:
Yeah. That's why we have the `rescue` below. Any error that is thrown would be caught there and logged.

qingling128 commented:
Feedback addressed. PTAL.

qingling128 commented:
THX!

igorpeshansky commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -76,7 +80,20 @@ module Constants
+      # The map between a resource type and expected subservice attributes.
+      SUBSERVICE_METADATA_ATTRIBUTES = \
+        [APPENGINE_CONSTANTS, CONTAINER_CONSTANTS, DATAPROC_CONSTANTS]
+        .map { |consts| [consts[:resource_type], consts[:metadata_attributes]] }
```
If you added `.to_set` here, you could have avoided it in the constants themselves...

qingling128 replied:
Addressed in the follow up PR: https://github.com/GoogleCloudPlatform/fluent-plugin-google-cloud/pull/148/commits/f3fbf25a37cf73dc44576d57be30ea77f7a0c322.

igorpeshansky commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -873,15 +657,398 @@ def fetch_gce_metadata(metadata_path)
+    rescue StandardError => e
+      @log.error "Failed to set monitored resource labels for #{type}: ",
+                 error: e
+      return {}
```
This could have been just `{}` as well...

qingling128 replied:
Addressed in the follow up PR: https://github.com/GoogleCloudPlatform/fluent-plugin-google-cloud/pull/148/commits/f3fbf25a37cf73dc44576d57be30ea77f7a0c322.

igorpeshansky commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -873,15 +657,398 @@ def fetch_gce_metadata(metadata_path)
+      @tag_regexp_list.each do |derived_type, tag_regexp|
+        matched_regex_group = tag_regexp.match(tag)
+        return [derived_type, matched_regex_group] if
+          matched_regex_group
```
Fits on the previous line.

qingling128 replied:
This no longer applies in the follow up PR: https://github.com/GoogleCloudPlatform/fluent-plugin-google-cloud/pull/148/commits/f3fbf25a37cf73dc44576d57be30ea77f7a0c322.

igorpeshansky commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -873,15 +657,398 @@ def fetch_gce_metadata(metadata_path)
+          resource_labels_candidates =
+            matched_regex_group.names.zip(matched_regex_group.captures).to_h
+          common_labels_candidates =
+            resource_labels_candidates.dup
```
Fits on the previous line.

qingling128 replied:
Addressed in the follow up PR: https://github.com/GoogleCloudPlatform/fluent-plugin-google-cloud/pull/148/commits/f3fbf25a37cf73dc44576d57be30ea77f7a0c322.

igorpeshansky commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -1178,43 +1338,35 @@ def decode_cloudfunctions_function_name(function_name)
         do |(original_label, new_label), extracted_labels|
-        extracted_labels[new_label] = convert_to_utf8(
-          record.delete(original_label).to_s) if record.key?(original_label)
+        extracted_labels[new_label] =
```
Alternatively:
```ruby
        value = hash.delete(original_label)
        extracted_labels[new_label] = convert_to_utf8(value.to_s) if value
```

qingling128 replied:
Addressed in the follow up PR: https://github.com/GoogleCloudPlatform/fluent-plugin-google-cloud/pull/148/commits/f3fbf25a37cf73dc44576d57be30ea77f7a0c322.
